File: .agents/retrievers/ast_neighbors.py
"""
Find files that reference symbols defined in seed files.
Uses ripgrep for speed. Returns a list of file paths.
"""
import json
import re
import subprocess
import sys
from pathlib import Path


SKIP_DIRS = ["node_modules", "vendor", "dist", "build", ".git", "__pycache__"]

SYMBOL_PATTERNS = {
    ".ts":  re.compile(r"""export\s+(?:default\s+)?(?:function|class|const|type|interface|enum)\s+(\w+)"""),
    ".tsx": re.compile(r"""export\s+(?:default\s+)?(?:function|class|const|type|interface|enum)\s+(\w+)"""),
    ".js":  re.compile(r"""export\s+(?:default\s+)?(?:function|class|const)\s+(\w+)"""),
    ".py":  re.compile(r"""^(?:def|class)\s+(\w+)""", re.MULTILINE),
    ".go":  re.compile(r"""^func\s+(?:\(\w+\s+\*?\w+\)\s+)?(\w+)""", re.MULTILINE),
}


def extract_symbols(path: Path) -> list[str]:
    pattern = SYMBOL_PATTERNS.get(path.suffix)
    if not pattern:
        return []
    try:
        content = path.read_text(errors="ignore")
    except OSError:
        return []
    return pattern.findall(content)


def find_references(symbols: list[str], root: str, exclude_files: list[str]) -> list[str]:
    if not symbols:
        return []

    rg_cmd = [
        "rg", "--files-with-matches", "--type-add", "code:*.{ts,tsx,js,jsx,py,go}",
        "--type", "code",
    ]
    for skip in SKIP_DIRS:
        rg_cmd += ["--glob", f"!{skip}"]
    rg_cmd += ["|".join(re.escape(s) for s in symbols), root]

    try:
        result = subprocess.run(rg_cmd, capture_output=True, text=True, timeout=30)
        files = result.stdout.strip().splitlines()
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return []

    exclude_set = {Path(f).resolve() for f in exclude_files}
    return [f for f in files if Path(f).resolve() not in exclude_set]


def neighbors(seed_files: list[str], root: str) -> list[str]:
    all_refs: set[str] = set()
    for seed in seed_files:
        path = Path(seed)
        if not path.exists():
            continue
        symbols = extract_symbols(path)
        refs = find_references(symbols, root, seed_files)
        all_refs.update(refs)
    return sorted(all_refs)


if __name__ == "__main__":
    seeds = sys.argv[1:-1]
    root = sys.argv[-1]
    result = neighbors(seeds, root)
    print(json.dumps(result, indent=2))

File: internal/core/ast_neighbors.go
package core

import (
	"bytes"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var astSkipDirs = []string{"node_modules", "vendor", "dist", "build", ".git", "__pycache__"}

var symbolPatterns = map[string]*regexp.Regexp{
	".ts":  regexp.MustCompile(`export\s+(?:default\s+)?(?:function|class|const|type|interface|enum)\s+(\w+)`),
	".tsx": regexp.MustCompile(`export\s+(?:default\s+)?(?:function|class|const|type|interface|enum)\s+(\w+)`),
	".js":  regexp.MustCompile(`export\s+(?:default\s+)?(?:function|class|const)\s+(\w+)`),
	".py":  regexp.MustCompile(`(?m)^(?:def|class)\s+(\w+)`),
	".go":  regexp.MustCompile(`(?m)^func\s+(?:\(\w+\s+\*?\w+\)\s+)?(\w+)`),
}

func extractSymbols(path string) []string {
	ext := filepath.Ext(path)
	pattern, ok := symbolPatterns[ext]
	if !ok {
		return nil
	}
	content, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	matches := pattern.FindAllSubmatch(content, -1)
	var symbols []string
	for _, match := range matches {
		if len(match) > 1 {
			symbols = append(symbols, string(match[1]))
		}
	}
	return symbols
}

func findReferences(symbols []string, root string, excludeFiles []string) []string {
	if len(symbols) == 0 {
		return nil
	}
	args := []string{"--files-with-matches", "--type-add", "code:*.{ts,tsx,js,jsx,py,go}", "--type", "code"}
	for _, skip := range astSkipDirs {
		args = append(args, "--glob", "!"+skip)
	}
	
	var escapedSymbols []string
	for _, s := range symbols {
		escapedSymbols = append(escapedSymbols, regexp.QuoteMeta(s))
	}
	args = append(args, strings.Join(escapedSymbols, "|"), root)

	cmd := exec.Command("rg", args...)
	var stdout bytes.Buffer
	cmd.Stdout = &stdout
	err := cmd.Run()
	if err != nil {
		return nil
	}

	excludeSet := make(map[string]bool)
	for _, f := range excludeFiles {
		abs, err := filepath.Abs(f)
		if err == nil {
			excludeSet[filepath.ToSlash(abs)] = true
		}
	}

	var files []string
	for _, line := range strings.Split(stdout.String(), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		abs, err := filepath.Abs(line)
		if err == nil {
			if !excludeSet[filepath.ToSlash(abs)] {
				files = append(files, line)
			}
		}
	}
	return files
}

func astNeighbors(seedFiles []string, root string) []string {
	allRefs := make(map[string]bool)
	for _, seed := range seedFiles {
		if _, err := os.Stat(seed); err != nil {
			continue
		}
		symbols := extractSymbols(seed)
		refs := findReferences(symbols, root, seedFiles)
		for _, ref := range refs {
			allRefs[ref] = true
		}
	}
	var res []string
	for r := range allRefs {
		res = append(res, r)
	}
	sort.Strings(res)
	return res
}
